### PR TITLE
scripts: kconfig: Replace conf with Python

### LINKED ISSUE #5374: merge_config.sh can behave differently from merge_config.py
Steps to reproduce:

1. Change the buildscripts to use merge_config.py.
2. ./scripts/sanitycheck -Tsamples/drivers/led_ws2812/

The sample fails to build with merge_config.py.

### LINKED ISSUE #5419: Provide a python based kconfig processing script, replacement for conf/mconf..
Using kconfiglib, provide a python based conf replacement to allow native building on windows and drop dependency on host compiler.

### PR DESCRIPTION
**Pull request by carlescufi**
fixes #5374
fixes #5419

### CODE REVIEW

**Comment by carlescufi** [comment]
This completely replaces conf and merge_config with a single kconfig.py Python script.

File: cmake/kconfig.cmake
Comment (carlescufi) [inline cmake/kconfig.cmake]: This is a leftover from debugging, to check that the `kconfig/` folder, which is no longer being generated, is not required. Can be removed.

File: scripts/kconfig/kconfig.py
Comment (carlescufi) [inline scripts/kconfig/kconfig.py]: We can replace the header of the `.config` file if we want to

File: scripts/kconfig/kconfig.py
Comment (carlescufi) [inline scripts/kconfig/kconfig.py]: We can replace the header of the `autoconf.h` file if we want to

File: scripts/kconfig/kconfiglib.py
Comment (carlescufi) [inline scripts/kconfig/kconfiglib.py]: This is equivalent to an MIT license: https://en.wikipedia.org/wiki/ISC_license

**Comment by nashif** [comment]
you probably want to mention in the kconfiglib commit that it will replace the current version in doc. So it is not something new that need to be approved.

**Comment by SebastianBoe** [comment]
When building hello_world with BOARD=nrf52_pca10040

> sebo@mach:~/zephyr/samples/hello_world$ cb
> -- Found PythonInterp: /usr/bin/python3 (found suitable version "3.5.2", minimum required is "3.4") 
> -- Selected BOARD nrf52_pca10040
> Zephyr version: 1.10.99
> warning: COVERAGE was assigned the value "y" but got the value "n" -- check dependencies
> -- Generating zephyr/include/generated/generated_dts_board.h
> -- The C compiler identification is GNU 6.2.0
> 

Can we improve the UX here, make it more obvious that the warning is coming from Kconfig for example?

**Comment by SebastianBoe** [comment]
I liked that the old system listed which files were used to merge configs,
I don't think we should remove that.

> Using /home/sebo/zephyr/boards/arm/nrf52_pca10040/nrf52_pca10040_defconfig as base
> Merging /home/sebo/zephyr/samples/hello_world/prj.conf
> Merging /home/sebo/zephyr/tests/include/test.config
>

**Comment by carlescufi** [comment]
Note there's no support for `menuconfig` or any other graphical configuration system yet.  I want to concentrate on the sanitycheck first.

**Comment by SebastianBoe** [comment]
Ah, ok, please mark as WIP/RFC.

**Comment by SebastianBoe** [comment]
CMake clean configure time regressed significantly :(

0.76 seconds in master
0.97 seconds now

It seems that it is doing more, which can give us a better UX, so 50ms or so performance degradation would be OK. but 200ms is quite a long time IMHO ...

File: cmake/kconfig.cmake
Comment (SebastianBoe) [reply cmake/kconfig.cmake]: +1 to removing it.

**Comment by carlescufi** [comment]
> 0.76 seconds in master
> 0.97 seconds now

I'm sure this is true for Linux. But how about the same on MSYS2 vs running this on a native Windows prompt with a native Python build?

**Comment by SebastianBoe** [comment]
> I'm sure this is true for Linux. But how about the same on MSYS2 vs running this on a native Windows prompt with a native Python build?

I don't know ;) I still don't understand why msys runs so slow.

**Comment by carlescufi** [comment]
Additional info: I compared the `.config` and `autconf.h` files generated by `conf` and `kconfig.py`. While the 2 `.config` are identical, Kconfiglib adds a `#define CONFIG_KERNELVERSION` to `autoconf.h` that was not present in the version generated by `conf`. I doubt this is an issue but I wanna raise it here in case.

**Comment by SebastianBoe** [comment]
This used to work, and should continue to do so:

cmake ..
# Edit a Kconfig file, e.g. zephyr/misc/Kconfig
cmake ..

Expected result:

Kconfig outputs will reflect Kconfig inputs (Such as Kconfig
files and .config files)

Intuitively; Kconfig is a part of configuration, and configuration is done in CMake's configure stage, so
running the configure stage (cmake .. or make rebuild_cache) should result in a configuration that is synced with the sources.

**Comment by SebastianBoe** [comment]
About KERNELVERSION. The user is supposed to be able to assume, that all # CONFIG_FOO_BAR defines he magically sees as available in his code, should also be available in CMake build scripts.

e.g. he should be able to do

if(CONFIG_KERNELVERSION STREQUAL bla bla)

But I see that CONFIG_KERNELVERSION is not entered into the CMake namespace (because it is not in the .config file), this should be fixed.

**Comment by nashif** [comment]
for the kconfiglib import, it would be cleaned if you import the upstream version and have another patch with the zephyr related changes.

**Comment by carlescufi** [comment]
@nashif I did that initially actually. But then I thought it was better to point to my GitHub repo instead. Maybe we could keep Kconfiglib in a Zephyr-owned repo? The reason is that Zephyr patches != Kconfiglib patches, and it's a pain to  maintain later.

**Comment by carlescufi** [comment]
@nashif @SebastianBoe Addressed comments except for the dependency problem, which @SebastianBoe is looking at.

**Comment by carlescufi** [comment]
@SebastianBoe 

>But I see that CONFIG_KERNELVERSION is not entered into the CMake namespace (because it is not in > the .config file), this should be fixed.

This is not new, `conf` did not store `CONFIG_KERNELVERSION` into the `.config` file either (although it did not store it in the `autoconf.h` neither). I think this is minor and it also seems to be the case for Linux from what I can gather.

**Comment by SebastianBoe** [comment]
> I think this is minor and it also seems to be the case for Linux from what I can gather. I've actually sent an email to @ulfalizer asking exactly this, to clarify.

I disagree that it is minor. But it doesn't need to block the merge as far as I am concerned. We can just open a bug for it and fix it later if you want.

**Comment by carlescufi** [comment]
>I disagree that it is minor. But it doesn't need to block the merge as far as I am concerned. We can just >open a bug for it and fix it later if you want.

OK, I think I will do that unless @ulfalizer confirms this is an issue in Kconfiglib that needs fixing, in which case I will try to fix it upstream.

**Comment by carlescufi** [comment]
> Can we improve the UX here, make it more obvious that the warning is coming from Kconfig for 
> example?

What I could do is print a line just before I construct the `Kconfig` object so that we know that output after that is part of the Kconfig tree parsing. What do you think?

File: scripts/kconfig/kconfig.py
Comment (mbolivar) [inline scripts/kconfig/kconfig.py]: And thus ISC license as well? Very small script so perhaps less of a concern?

File: scripts/kconfig/kconfig.py
Comment (mbolivar) [inline scripts/kconfig/kconfig.py]: Just to keep things tidy and reviewable since this is not minimal, I think putting kconfiglib.py in ext/ (one patch to add the upstream copy, another with your changes) would be appropriate.

You can explicitly add `os.path.join(os.path.abspath(os.environ['ZEPHYR_BASE']), 'ext', 'subdir', 'containing', 'kconfiglib')` to sys.path before performing the import to ensure it's visible as done here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/conf.py#L21

**Review by mbolivar — COMMENTED** [review]
Hurray for not reinventing the wheel!

Just some license/location concerns, otherwise I think this approach is good.

File: scripts/kconfig/kconfig.py
Comment (carlescufi) [reply scripts/kconfig/kconfig.py]: Yes, ISC as well, so that is not an issue.

File: scripts/kconfig/kconfig.py
Comment (carlescufi) [reply scripts/kconfig/kconfig.py]: I am not sure about that because `ext/` is supposed to hold *unmodified copies* of external repos. Instead I suggest we fix our Kconfig structure and then use `pip` to install Kconfiglib at a later date.

Comment (carlescufi) [reply scripts/kconfig/kconfig.py]: @nashif can you weigh in?

Comment (nashif) [reply scripts/kconfig/kconfig.py]: ext/ is for stuff that ends up in the binary, external components and middleware, we never put original  kconfig in ext/ and we also never put kbuild, kconfig is part of the build and configuration system and IMO should be in scripts/ just like it used to be before.

Comment (nashif) [reply scripts/kconfig/kconfig.py]: and the other thing, we want to get rid of this eventually anyways and use the upstream version, so it will go away sooner than later.

Comment (mbolivar) [reply scripts/kconfig/kconfig.py]: > ext/ is supposed to hold unmodified copies of external repos

That's not correct across the board. For example, STM32 HALs have received patches, mostly to fix upstream bugs:

0329450088fdb1ff9c6e31011496b907283cd950
9f420955d9a3cca4573a2a82d36e0e7b92cdcf3b

Looks like Nordic HALs are altered as well:

fc78ddf7d3a20d99669e49c0aba9ea1c9b8c702f
71c3d340da8cc43afb019d31756b818298ccddb9

The policy has been as I described in my experience. In the ST case, there are even README files tracking the patch list. I thought this was officially documented but can't find it right now

Comment (mbolivar) [reply scripts/kconfig/kconfig.py]: Commit template for external components:
http://docs.zephyrproject.org/contribute/contribute_guidelines.html#identifying-contribution-origin

Requirement that non-Apache lives in ext/:
http://docs.zephyrproject.org/contribute/contribute_non-apache.html

Still can't find documentation about splitting up upstream import from Zephyr patches, but I'm sure that's generally what's expected

Comment (mbolivar) [reply scripts/kconfig/kconfig.py]: > ext/ is for stuff that ends up in the binary, external components and middleware, we never put original kconfig in ext/ and we also never put kbuild, kconfig is part of the build and configuration system and IMO should be in scripts/ just like it used to be before.

Okay. Should I send patches to the documentation to make this more clear?

Comment (carlescufi) [reply scripts/kconfig/kconfig.py]: In general you raise good points @mbolivar, but I still think `kconfiglib.py` should live in scripts/kconfig.

> That's not correct across the board. 

You are right, and the Nordic ones are definitely incorrect oversights that will be fixed soon once we replace what we have now with `nrfx`. 

> Commit template for external components:

Yes, I did not follow that strictly because I assumed that was only valid for `ext/` :) bit of a catch-22 here.

> Requirement that non-Apache lives in ext/:

I think that this applies only to firmware, i.e. source code that once compiled ends up in the image to flash.

Comment (carlescufi) [reply scripts/kconfig/kconfig.py]: FWIW @mbolivar we even have GPL code in `scripts/kconfig`! https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/kconfig/gconf.c#L5

**Comment by carlescufi** [comment]
@mbolivar 

>Hurray for not reinventing the wheel!

Indeed! Credit goes to @ulfalizer for a great Python Kconfig library though, as well as all the help we've gotten from him so far.

File: scripts/kconfig/kconfig.py
Comment (nashif) [reply scripts/kconfig/kconfig.py]: kconfiglib.py ALREADY exists in the tree, this is just an update to an existing library we already have. We did not remove the old one because there is another script that depend on it for generating docs, but this should be done in parallel and be ready before this PR is merged. I started looking at that.

Comment (mbolivar) [reply scripts/kconfig/kconfig.py]: OK OK, my mistake :).

I'm just going to send a docs patch to try to clear this up, we can discuss that there

**Comment by carlescufi** [comment]
>I disagree that it is minor. But it doesn't need to block the merge as far as I am concerned. We can just >open a bug for it and fix it later if you want.

No need. @ulfalizer has fixed it upstream and I've now updated to the latest Kconfiglib.

File: cmake/kconfig.cmake
Comment (carlescufi) [inline cmake/kconfig.cmake]: @SebastianBoe does this look OK to you?

File: cmake/host-tools.cmake
Comment (carlescufi) [inline cmake/host-tools.cmake]: Keep looking for `conf` in non-Windows builds, including MSYS2.

**Comment by carlescufi** [comment]
@nashif @SebastianBoe I now tested it on Windows' `cmd.exe`, runs fine.

File: cmake/kconfig.cmake
Comment (SebastianBoe) [inline cmake/kconfig.cmake]: s/run/${kconfig_target}/

File: cmake/kconfig.cmake
Comment (SebastianBoe) [reply cmake/kconfig.cmake]: LGTM

**Comment by ulfalizer** [comment]
Hello,

Glad it's helpful!

I'm currently looking for a job by the way. If you want me to build a good menuconfig for you or implement other more advanced Kconfig- or build-related features needed in Zephyr, then I'm available for freelance work.

I'm always happy to help if you decide to do it yourselves. Just putting it out there as an option. :)

Cheers,
Ulf

**Comment by carlescufi** [comment]
@ulfalizer 

>I'm currently looking for a job by the way. If you want me to build a good menuconfig for you or implement other more advanced Kconfig- or build-related features needed in Zephyr, then I'm available for freelance work.

Thanks for letting us know, I will have a talk internally and get back to you privately.

**Comment by carlescufi** [comment]
This is ready to be merged as far as @SebastianBoe and myself are concerned.

**Comment by nashif** [comment]
will give it another try today. AFAIK, it looks great already. Need to remove the old Kconfiglib from the tree at the same time.

**Comment by mbolivar** [comment]
> This is ready to be merged as far as @SebastianBoe and myself are concerned.

+1

**Review by SebastianBoe — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
